apm-agent-core/src/test/java/co/elastic/apm/agent/impl/circuitbreaker/CircuitBreakerTest.java

```diff
@@ -74,104 +97,237 @@ public void tearDown() {
+        // stress pauses when recording enabled
+        // timeline   1  2  3  4  5
+        // stress     ---sssssss-----
+        // recording  xxxxxx------xxx
+        // state      rrr---------rrr
+
+        // 1
+        circuitBreakerConfiguration.circuitBreakerEnabled.set(true);
+        assertRunning();
+
+        // 2
+        monitor.simulateStress();
+        awaitPaused();
+
+        // 3 recording = false under stress should not change state
         TracerInternalApiUtils.setRecordingConfig(config, false, TEST_CONFIG_SOURCE_NAME);
-        pollCount = monitor.simulateStressRelieved();
-        monitor.waitUntilPollCounterIsGreaterThan(pollCount + 1);
-        // should still be PAUSED as the state is inactive
-        assertThat(tracer.getState()).isEqualTo(PAUSED);
+        assertState(this::assertPaused);
+
+        // 4 stress ends, should still be paused due to recording = false
+        monitor.simulateStressRelieved();
+        assertState(this::assertPaused);
+
+        // 5
+        // configuration recording = true should make it run again
         TracerInternalApiUtils.setRecordingConfig(config, true, TEST_CONFIG_SOURCE_NAME);
-        assertThat(tracer.getState()).isEqualTo(RUNNING);
+        awaitRunning();
     }
 
     @Test
-    void testReactivateThenStressRelief() throws IOException {
-        doReturn(true).when(circuitBreakerConfiguration).isCircuitBreakerEnabled();
-        assertThat(tracer.getState()).isEqualTo(RUNNING);
+    void testPauseThroughConfigThenResumeOnlyWhenStressRelieved() throws IOException, InterruptedException {
+        // enable recording while under stress does not trigger runnable state
+        // timeline   1  2  3  4  5
+        // stress     ------ssssss----
+        // recording  xxx------xxxxxx
+        // state      rrr---------rrr
+
+        // 1
+        circuitBreakerConfiguration.circuitBreakerEnabled.set(true);
+        assertRunning();
+
+        // 2 recording = false should pause
         TracerInternalApiUtils.setRecordingConfig(config, false, TEST_CONFIG_SOURCE_NAME);
-        assertThat(tracer.getState()).isEqualTo(PAUSED);
+        awaitPaused();
+
+        // 3 stress should keep it paused
         monitor.simulateStress();
+        assertState(this::assertPaused);
+
+        // 4 should not resume tracer as we are under stress
         TracerInternalApiUtils.setRecordingConfig(config, true, TEST_CONFIG_SOURCE_NAME);
-        // check that reactivation now has no effect even after waiting for the next resume poll
-        int pollCount = monitor.getPollCount();
-        monitor.waitUntilPollCounterIsGreaterThan(pollCount + 1);
-        assertThat(tracer.getState()).isEqualTo(PAUSED);
-        // check that stress relief now reactivates
-        pollCount = monitor.simulateStressRelieved();
-        monitor.waitUntilPollCounterIsGreaterThan(pollCount + 1);
-        assertThat(tracer.getState()).isEqualTo(RUNNING);
+        assertState(this::assertPaused);
```

Comment:
You need to `awaitPoll` somehow. It doesn't make sense to assert that the state remained the same without that.